Sources/Data/models/RecentlyClosed.swift

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import CoreData
+import Shared
+import os.log
+
+public struct SavedRecentlyClosed {
+  public let url: String
+  public let title: String?
+  public let dateAdded: Date?
+  public let historyList: [String]
+  public let historyIndex: Int16
+
+  public init(url: String, title: String?, dateAdded: Date? = Date(), historyList: [String], historyIndex: Int16) {
+    self.url = url
+    self.title = title
+    self.dateAdded = dateAdded
+    self.historyList = historyList
+    self.historyIndex = historyIndex
+  }
+}
+
+public final class RecentlyClosed: NSManagedObject, CRUD {
+  @NSManaged public var url: String
+  @NSManaged public var title: String?
+  @NSManaged public var dateAdded: Date?
+  @NSManaged public var historyList: NSArray?  // List of urls for back forward navigation
+  @NSManaged public var historyIndex: Int16
+  
+  public class func get(with url: String) -> RecentlyClosed? {
+    return getInternal(with: url)
+  }
+
+  public class func all() -> [RecentlyClosed] {
+    let sortDescriptors = [NSSortDescriptor(key: #keyPath(RecentlyClosed.dateAdded), ascending: false)]
+    return all(sortDescriptors: sortDescriptors) ?? []
+  }
+
+  public class func remove(with url: String) {
+    DataController.perform { context in
+      if let item = getInternal(with: url, context: context) {
+        item.delete(context: .existing(context))
+      }
+    }
+  }
+
+  public static func removeAll() {
+    RecentlyClosed.deleteAll()
+  }
+  
+  public class func insert(_ saved: SavedRecentlyClosed) {
+    DataController.perform { context in
+      guard let entity = entity(in: context) else {
+        Logger.module.error("Error fetching the entity 'Recently Closed' from Managed Object-Model")
+  
+        return
+      }
+  
+      let source = RecentlyClosed(entity: entity, insertInto: context)
+      source.url = saved.url
+      source.title = saved.title
+      source.dateAdded = saved.dateAdded
+      source.historyList = saved.historyList as NSArray
+      source.historyIndex = saved.historyIndex
+    }
+  }
+  
+  public class func insertAll(_ savedList: [SavedRecentlyClosed]) {
+    savedList.forEach {
+      RecentlyClosed.insert($0)
```

Comment:
Yes. This WILL cause a performance issue on older devices.
It should do: `RecentlyClosed.insert(all: savedList)` or similar so it creates one context and inserts all on that context, and only calls `saveContext` once. Otherwise for every item it will call `saveContext` which is very costly!

Reply:
Appreciate the warning, check out the fixin https://github.com/brave/brave-ios/pull/6755/commits/1417d22c5c8aff13efc21f8c30cb3e79579859f5